api/src/services/users.js

```diff
@@ -532,48 +551,128 @@ const validateUserContact = (data, user, userSettings) => {
+const createUserEntities = async (user, appUrl, preservePrimaryContact) => {
   const response = {};
   await validateNewUsername(user.username);
   await createPlace(user);
   await setContactParent(user);
   await createContact(user, response);
-  await storeUpdatedPlace(user);
+  await storeUpdatedPlace(user, preservePrimaryContact);
   await createUser(user, response);
   await createUserSettings(user, response);
   await tokenLogin.manageTokenLogin(user, appUrl, response);
   return response;
 };
 
-const validateUserFields = (users) => {
-  const missingFieldsFailingIndexes = [];
-  const tokenLoginFailingIndexes = [];
-  const passwordFailingIndexes = [];
-  users.forEach((user, index) => {
-    const fields = missingFields(user);
-    const hasMissingFields = fields.length > 0;
-    if (hasMissingFields) {
-      missingFieldsFailingIndexes.push({ fields, index });
-      return;
-    }
+const assignCsvCellValue = (data, attribute, value) => {
+  attribute = (attribute || '').trim();
+  if (!attribute.length || attribute.toLowerCase().indexOf(':excluded') > 0) {
+    return;
+  }
+  data[attribute] = typeof value === 'string' ? value.replace(/^"|"$/g, '').trim() : value;
+};
 
-    const tokenLoginError = tokenLogin.validateTokenLogin(user, true);
-    if (tokenLoginError) {
-      tokenLoginFailingIndexes.push({ tokenLoginError, index });
-      return;
+const parseCsvRow = (data, header, value, valueIdx) => {
+  const deepAttributes = header[valueIdx].split('.');
+  if (deepAttributes.length === 1) {
+    assignCsvCellValue(data, deepAttributes[0], value);
+    return;
+  }
+
+  let deepObject = data;
+  deepAttributes.forEach((attr, idx) => {
+    if (idx === deepAttributes.length - 1) {
+      return assignCsvCellValue(deepObject, attr, value);
     }
 
-    const passwordError = validatePassword(user.password);
-    if (passwordError) {
-      passwordFailingIndexes.push({ passwordError, index });
+    if (!deepObject[attr]) {
+      assignCsvCellValue(deepObject, attr, {});
     }
+    deepObject = deepObject[attr];
   });
 
-  return {
-    missingFieldsFailingIndexes,
-    tokenLoginFailingIndexes,
-    passwordFailingIndexes,
+  return data;
+};
+
+const parseCsv = async (csv, logId) => {
+  if (!csv || !csv.length) {
+    throw new Error('CSV is empty.');
+  }
+
+  let allRows = csv.split(/\r\n|\n/);
+  const header = allRows
+    .shift()
+    .split(',');
+  allRows = allRows.filter(row => row.length);
+
+  const progress = {
+    status: BULK_UPLOAD_PROGRESS_STATUSES.PARSING,
+    parsing: { total: allRows.length, failed: 0, successful: 0 }
   };
+  const logData = [];
+  await bulkUploadLog.updateLog(logId, progress);
+
+  const users = [];
+  const ignoredUsers = {};
+  for (let rowIdx = 0; rowIdx < allRows.length; rowIdx++) {
+    let ignoreUser = false;
+    let ignoreMessage = '';
+    const user = {};
+    try {
+      allRows[rowIdx]
+        .split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/)
```

Comment:
At a minimum, this is probably a good case for a comment, to show what an expected string looks like. As with most comment scenarios though, some tests can tell the story better ;) 

A function that takes a single row and produces whatever split returns would help to contain the complexity with the regex here and then let you have a few test cases to make sure it stays working as expected over time.